src/benchmarklib/benchmark_runner.cpp

```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
   std::random_device random_device;
   std::mt19937 random_generator(random_device());
 
-  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
-  while (state.keep_running()) {
-    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
-
-    for (const auto& named_query : mutable_named_queries) {
-      const auto query_benchmark_begin = std::chrono::steady_clock::now();
-
-      // Execute the query, we don't care about the results
-      _execute_query(named_query);
-
-      const auto query_benchmark_end = std::chrono::steady_clock::now();
-
-      auto& query_benchmark_result = _query_results_by_query_name[named_query.first];
-      query_benchmark_result.duration += query_benchmark_end - query_benchmark_begin;
-      query_benchmark_result.num_iterations++;
+  const auto number_of_queries_per_set = mutable_named_queries.size();
+
+  // The atomic uints are modified by other threads when finishing a query set, to keep track of when we can
+  // let a simulated client schedule the next set, as well as the total number of finished query sets so far
+  auto currently_running_query_sets = std::atomic_uint{0};
+  auto finished_query_set_runs = std::atomic_uint{0};
+  auto finished_queries_total = std::atomic_uint{0};
+  // For the iteration durations vector, we need a mutex
+  auto result_mutex = std::mutex{};
+
+  auto tasks = std::vector<std::shared_ptr<AbstractTask>>{};
+  auto state = BenchmarkState{_config.max_num_query_runs, _config.max_duration};
+
+  while (state.keep_running(finished_query_set_runs.load(std::memory_order_relaxed))) {
```

Comment:
I don't understand this line

Reply:
Which part?
1) `state.keep_running()` now takes a parameter of how many query (sets) are already finished (will document that in the BenchmarkState declaration)
2) `.load()` is to get the actual value from std::atomic
3) `std::memory_order_relaxed` tells `load` that there are no synchronization or ordering constraints imposed on other reads or writes, only this operation's atomicity is guaranteed. Which, according to [this](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering), should be fine here. However after thinking more about it now, I will reevaluate the memory_ordering uses in the other cases as well.

Reply:
The first. Even with the documentation, I think that the signature of that method is everything but easy to understand. Why do I have to pass a number into the state to find out if the benchmark should continue? Can't we do the check outside of `keep_running`, i.e.

`while(state.keep_running() && ...`?

Reply:
Yes we can. This would leave the BenchmarkState as a timer basically.